edns_test.go

```diff
@@ -62,7 +62,7 @@ func TestOPTTtl(t *testing.T) {
 	}
 
 	e.SetExtendedRcode(42)
-	if e.ExtendedRcode() != 42 {
-		t.Errorf("set 42, expected %d, got %d", 42-15, e.ExtendedRcode())
+	if e.ExtendedRcode() != (42 & 0xFF0) {
```

Comment:
Why doesn't ExtendedRcode return 42?

Reply:
The original code stores `v - 15` into OPT RR.
The correct behavior is to store `v >> 4` into OPT RR.

The currently only available extended RCODE (`RcodeBadVers = 16`), produces the coincidence that `16 - 15 == 16 >> 4`, therefore the existing code is not causing problem with **real-life** DNS packets at present.

By adapting the correct behavior, I need to discard the lower 4 bits from `v` to make sure the function still behaves correctly with `e.SetExtendedRcode(RcodeBadVers)`.
Therefore the discarded lower 4 bits is intended to cause an input of 42 to produce an output of 32.

Another method to fix this problem is to ask the caller to discard the lower 4 bits themselves before calling `SetExtendedRcode`, making the output still 42. But this might break any existing third-party code.

---

By the way, `msg.go` is already using `SetExtendedRcode` [the wrong way](https://github.com/miekg/dns/blob/e4205768578dc90c2669e75a2f8a8bf77e3083a4/msg.go#L708). It discarded the lower 4 bits **before** calling, resulting `SetExtendedRcode` always falling into the empty `return` statement. I have also modified it to use the consistent behavior.

I have also updated the docstring, and intentionally changed the function signature to bring in incompatibility. I think this change will not introduce problem to third-party code without being noticed.

Reply:
What I meant; if we properly fix this (by dropping `uint8`) this should just return 42?

Reply:
> What I meant; if we properly fix this (by dropping `uint8`) this should just return 42?

No I don't think so.

Extended Rcode only records the upper 8 bits. The lower 4 bits is still in the old place.
You should either drop the last 4 bits yourself, or ask the caller to drop the 4 bits for you.
For the former, 42 should not just return 42. For the latter, existing users might be confused.